### ISSUE TITLE
[REVIEW]: DBSP_DRP: A Python package for automated spectroscopic data reduction of DBSP data

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3612]
**Submitting author:** @finagle29 (<a href="http://orcid.org/0000-0003-1118-3132">Milan Roberson</a>)
**Repository:** https://github.com/finagle29/DBSP_DRP/
**Branch with paper.md** (empty if default branch): 
**Version:** v1.0.0
**Editor:** @arfon
**Reviewers:** @crhea93, @arjunsavel
**Archive:** 10.5281/zenodo.6241526

 

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/8d33523aaa8283d0547994eb51b4407f/status.svg)](https://joss.theoj.org/papers/8d33523aaa8283d0547994eb51b4407f)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/8d33523aaa8283d0547994eb51b4407f"><img src="https://joss.theoj.org/papers/8d33523aaa8283d0547994eb51b4407f/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/8d33523aaa8283d0547994eb51b4407f/status.svg)](https://joss.theoj.org/papers/8d33523aaa8283d0547994eb51b4407f)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/finagle29/DBSP_DRP/" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@crhea93 & @arjunsavel, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @arfon know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @crhea93
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/finagle29/DBSP_DRP/">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/finagle29/DBSP_DRP/">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@finagle29) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @arjunsavel
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/finagle29/DBSP_DRP/">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/finagle29/DBSP_DRP/">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@finagle29) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @crhea93, @arjunsavel it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Wordcount for `paper.md` is 1158

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.12 s (496.0 files/s, 40014.8 lines/s)
--------------------------------------------------------------------------------
Language                      files          blank        comment           code
--------------------------------------------------------------------------------
Python                           28            582            427           2333
reStructuredText                 11            163            212            246
Markdown                          2             31              0            197
YAML                              6             29             11            166
TeX                               1              0              0            119
Bourne Shell                      3             16              8             42
CSS                               2              4              0             34
DOS Batch                         1              8              1             26
Dockerfile                        1             11             11             26
JavaScript                        1              3              0             13
make                              1              4              7              9
Bourne Again Shell                1              0              7              7
TOML                              1              0              0              7
--------------------------------------------------------------------------------
SUM:                             59            851            684           3225
--------------------------------------------------------------------------------


Statistical information for the repository 'b6da3713d45f98e5f691e63c' was
gathered on 2021/08/13.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Milan Roberson                 184          5523           2220          100.00

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Milan Roberson             3303           59.8          5.2                8.39
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.21105/joss.02308 is OK
- 10.1086/131027 is OK
- 10.3847/1538-4357/ab8943 is OK
- 10.3847/1538-4357/abbd98 is OK
- 10.3847/1538-4357/abaeec is OK
- 10.3847/1538-4357/abb45c is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03612/joss.03612/10.21105.joss.03612.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03612/joss.03612/10.21105.joss.03612.pdf) :page_facing_up: :point_left:

**Comment by arfon** [comment]
@crhea93, @arjunsavel – This is the review thread for the paper. All of our communications will happen here from now on.

Please read the "Reviewer instructions & questions" in the first comment above.

Both reviewers have checklists at the top of this thread (in that first comment) with the JOSS requirements. As you go over the submission, please check any items that you feel have been satisfied. There are also links to the JOSS reviewer guidelines.

The JOSS review is different from most other journals. Our goal is to work with the authors to help them meet our criteria instead of merely passing judgment on the submission. As such, the reviewers are encouraged to submit issues and pull requests on the software repository. When doing so, please mention `https://github.com/openjournals/joss-reviews/issues/3612` so that a link is created to this thread (and I can keep an eye on what is happening). Please also feel free to comment and ask questions on this thread. In my experience, it is better to post comments/questions/suggestions as you come across them instead of waiting until you've reviewed the entire package.

We aim for the review process to be completed within about 4-6 weeks but please make a start well ahead of this as JOSS reviews are by their nature iterative and any early feedback you may be able to provide to the author will be very helpful in meeting this schedule.

**Comment by crhea93** [comment]
This is a very interesting-looking package! While the majority of the documentation is present, I think it would benefit from 3 sections that are currently missing:
 - Functionality API
 - Automated Testing
 - Contributing Guidelines

**Comment by arjunsavel** [comment]
I’ve completed my review!

Overall:
The contributors to DBSP_DRP have produced an extensive and useful data reduction pipeline. The degree of its automation is impressive, and the contributors’ effort is evident — and there is clear scientific merit, with application to a well-known instrument on a well-known telescope. As someone who’s been on observing runs (though I wouldn’t call myself an observer!) and worked on a reduction pipeline, I’m particularly interested in this code's quicklook capability this code has.

I've opened a few issues in the repository. Many of the items I've mentioned are related to grammar and consistency.

- https://github.com/finagle29/DBSP_DRP/issues/16
- https://github.com/finagle29/DBSP_DRP/issues/15
- https://github.com/finagle29/DBSP_DRP/issues/14
- https://github.com/finagle29/DBSP_DRP/issues/13

I also wanted to touch base about the authorship list — I only see the first author in the commit history.

The various installation routes worked well for me. However, I couldn't fully verify the code's functionality — when I tried to reduce the sample_data, the fact that the data didn’t have RA / Dec ended ups throwing an Astropy units type error. Is there different data I should use to verify that this works, or am I missing something?

In summary, this package looks great to me. I’ll sign off for acceptance pending the resolution of the above issues!

**Comment by arfon** [comment]
@crhea93, @arjunsavel – thanks for your feedback here :zap:! @finagle29 – please let us know when you've had a chance to respond to the reviewer feedback.

**Comment by arfon** [comment]
:wave: @finagle29 – just checking in here to see how you're getting on responding to the reviewer feedback?

**Comment by finagle29** [comment]
Thanks for checking in!

I've been slowly working through the review issues that @arjunsavel opened. I've been a bit stuck on how to write contributing guidelines. Regarding the authorship list, Christoffer Fremling and Mansi Kasliwal were both substantial non-code contributors. The sample_data folder is slightly misnamed, it's intentionally broken FITS files that the unit tests rely on. I have put sample data to verify functionality here: https://drive.google.com/drive/folders/1HcbmItr50IN2D4GF37ycoZHa1Cco__6m?usp=sharing

I have two quick clarifying question for @crhea93. By functionality API, do you mean something like this automatically generated API documentation https://dbsp-drp.readthedocs.io/en/develop/dbsp_drp.html ? Also, could you elaborate on what Automated Testing documentation you would like to see?

**Comment by arjunsavel** [comment]
Hi @finagle29 — thank you for addressing my issues, things are looking great! I've closed all the completed issues.

With regard to community guidelines, you could use the emcee package's [`CONTRIBUTING.md`](https://github.com/dfm/emcee/blob/main/CONTRIBUTING.md) file as an example. In short, you'll need sections describing how to contribute to your package, how to report issues/bugs, and how to get help. Let me know if you have any specific questions about these!

Also, thanks for pointing me to some sample data — I'll be able to check that out in the next few days.

**Comment by arjunsavel** [comment]
Hi @finagle29 — sorry for the delay! 

I've been able to run the sample_data that you provided for the blue arm, and everything looks great there! However, I'm not able to do so for the red arm; in this case, I run into a consistent issue with each file (no RA, Dec --> astropy throws a TypeError when trying to create SkyCoords).

**Comment by finagle29** [comment]
No worries, @arjunsavel !

I ran into that bug myself, and have just pushed a fix to the `develop` branch. If this error (or other errors) persists, we can take discussion over to https://github.com/finagle29/DBSP_DRP/discussions.

**Comment by finagle29** [comment]
> I have two quick clarifying question for @crhea93. By functionality API, do you mean something like this automatically generated API documentation https://dbsp-drp.readthedocs.io/en/develop/dbsp_drp.html ? Also, could you elaborate on what Automated Testing documentation you would like to see?

Hi @crhea93, I just wanted to follow up on these questions I had regarding your comments

**Comment by finagle29** [comment]
> > And a question for @arfon : I have changed institutional affiliations since completing the vast majority of my work on DBSP_DRP, should I use my previous affiliation on the paper? If so, is there some way I should note my current affiliation?
> 
> You can include multiple affiliations if you wish? Perhaps listing your current and previous ones? The [example paper](https://joss.readthedocs.io/en/latest/submitting.html#example-paper-and-bibliography) shows how to do this.

Thanks for the example! I decided on listing my present address/affiliation in a footnote.

**Comment by arfon** [comment]
Happy new year folks! @arjunsavel – it looks like you're finished with your review, and at this point we're waiting on @crhea93 to complete their review based on @finagle29's updates. Is that a reasonable assessment of where we're at?

**Comment by arjunsavel** [comment]
Happy new year! @arfon, I've just completed going over some bug fixes that @finagle made — everything's looking great, and I'd consider my review completed.

**Comment by arfon** [comment]
Wonderful, thanks for the update @arjunsavel. @crhea93 – could you provide an update here on the status of your review?

**Comment by arfon** [comment]
Quick update: I just contacted @crhea93 over email to ask them to check back here.

**Comment by crhea93** [comment]
I'm sorry I don't know why I missed all of these messages! My apologies! I can say that I highly recommend the article.

**Comment by arfon** [comment]
> I'm sorry I don't know why I missed all of these messages! My apologies! I can say that I highly recommend the article.
 
 ⚡ , thanks for the update!

**Comment by arfon** [comment]
@finagle29 – At this point could you make a new release of this software that includes the changes that have resulted from this review. Then, please make an archive of the software in Zenodo/figshare/other service and update this thread with the DOI of the archive? For the Zenodo/figshare archive, please make sure that:

- The title of the archive is the same as the JOSS paper title
- That the authors of the archive are the same as the JOSS paper authors

I can then move forward with accepting the submission.

**Comment by finagle29** [comment]
Hi @arfon, I've created a new release including the changes from this review and archived it in Zenodo. The URL that points to this specific version (version 1.0.0) is https://doi.org/10.5281/zenodo.6241526, and https://doi.org/10.5281/zenodo.6241525 points to the latest version, I'm unsure which is better to use here.

**Comment by arfon** [comment]
@editorialbot set v1.0.0 as version

**Comment by arfon** [comment]
@editorialbot set 10.5281/zenodo.6241526 as archive

**Comment by arfon** [comment]
@editorialbot accept

**Comment by arfon** [comment]
@crhea93, @arjunsavel – many thanks for your reviews here! JOSS relies upon the volunteer effort of people like you and we simply wouldn't be able to do this without you ✨ 

@finagle29 – your paper is now accepted and published in JOSS :zap::rocket::boom:

**Comment by finagle29** [comment]
I just noticed a small issue: in the published version of the paper, there isn't a footnote after my name noting my present address/affiliation. When I use whedon to preview the paper on the `main` branch of the repo, the footnote is present.

**Comment by arfon** [comment]
I wonder if we're stripping this? As [per the docs](https://joss.readthedocs.io/en/latest/submitting.html#example-paper-and-bibliography), the preferred way would be to write this as such:

```
authors:
  - name: Milan S. Roberson^[Present address: Department of Physics and Astronomy, University of California, Los Angeles.]
    orcid: 0000-0003-1118-3132
    affiliation: "1,2"
  - name: Christoffer Fremling
    orcid: 0000-0002-4223-103X
    affiliation: 2
  - name: Mansi M. Kasliwal
    orcid: 0000-0002-5619-4938
    affiliation: 2
affiliations:
  - name: Schmidt Academy of Software Engineering, California Institute of Technology
    index: 1
  - name: Division of Physics, Mathematics and Astronomy, California Institute of Technology
    index: 2
date: 26 October 2021
bibliography: paper.bib
nocite: |
  @Lunnan2020
---
```

**Comment by tarleb** [comment]
The note should now be rendered in the PDF in the same way it did before. I'm now checking how the note can be included in JATS output.

**Comment by finagle29** [comment]
Thanks @tarleb !!
Is it possible to rerender the PDF displayed at https://joss.theoj.org/papers/10.21105/joss.03612 ?

**Comment by tarleb** [comment]
I believe @arfon might be able to do that.

**Comment by finagle29** [comment]
@arfon could you rerender the published PDF?
I noticed this just now, but as a result of the footnote issue, the citation string on the JOSS website includes a open square bracket as the last character of my last name:
`Roberson[ et al., (2022). DBSP_DRP: A Python package for automated spectroscopic data reduction of DBSP data. Journal of Open Source Software, 7(70), 3612, https://doi.org/10.21105/joss.03612`

**Comment by arfon** [comment]
@editorialbot accept

**Comment by arfon** [comment]
> @arfon could you rerender the published PDF?

This looks to be fixed now @finagle29?

**Comment by finagle29** [comment]
Yes, everything looks great, thank you so much for your help!

**Comment by arfon** [comment]
@editorialbot reaccept

**Comment by arfon** [comment]
@finagle29 – can you check this new PDF looks good? https://github.com/openjournals/joss-papers/blob/13c81ea7f6403838a32bbfbc9ae0835754b5e7d9/joss.03612/10.21105.joss.03612.pdf

**Comment by finagle29** [comment]
lgtm!
